File: src/main/kotlin/org/skialin/Pixmap.kt
package org.skialin

import org.skialin.impl.Managed
import org.skialin.impl.NativeLoader
import java.nio.ByteBuffer

class Pixmap private constructor(
    ptr: Long,
    private val keepAlive: Any?,
) : Managed(ptr, PixmapNative::nRelease) {
    val addr: Long get() = PixmapNative.nAddr(nativePtr)
    val rowBytes: Long get() = PixmapNative.nRowBytes(nativePtr)
    val width: Int get() = PixmapNative.nWidth(nativePtr)
    val height: Int get() = PixmapNative.nHeight(nativePtr)
    val isEmpty: Boolean get() = PixmapNative.nIsEmpty(nativePtr)
    val colorType: ColorType get() = ColorType.entries[PixmapNative.nColorType(nativePtr)]
    val alphaType: AlphaType get() = AlphaType.entries[PixmapNative.nAlphaType(nativePtr)]
    val isOpaque: Boolean get() = PixmapNative.nIsOpaque(nativePtr)
    val rowBytesAsPixels: Int get() = PixmapNative.nRowBytesAsPixels(nativePtr)
    val shiftPerPixel: Int get() = PixmapNative.nShiftPerPixel(nativePtr)

    fun computeByteSize(): Long = PixmapNative.nComputeByteSize(nativePtr)

    fun getColor(
        x: Int,
        y: Int,
    ): Color = PixmapNative.nGetColor(nativePtr, x, y)

    fun getAlphaf(
        x: Int,
        y: Int,
    ): Float = PixmapNative.nGetAlphaf(nativePtr, x, y)

    fun extractSubset(area: IRect): Pixmap? {
        val ptr = PixmapNative.nExtractSubset(nativePtr, area.left, area.top, area.right, area.bottom)
        return if (ptr == 0L) null else Pixmap(ptr, keepAlive)
    }

    fun readPixels(
        dst: Pixmap,
        srcX: Int = 0,
        srcY: Int = 0,
    ): Boolean = PixmapNative.nReadPixels(nativePtr, dst.nativePtr, srcX, srcY)

    fun scalePixels(
        dst: Pixmap,
        sampling: SamplingOptions = SamplingOptions.NEAREST,
    ): Boolean =
        PixmapNative.nScalePixels(
            nativePtr,
            dst.nativePtr,
            sampling.maxAniso,
            sampling.useCubic,
            sampling.cubicB ?: 0f,
            sampling.cubicC ?: 0f,
            sampling.filter.ordinal,
            sampling.mipmap.ordinal,
        )

    companion object {
        fun make(
            info: ImageInfo,
            buffer: ByteBuffer,
            rowBytes: Long,
        ): Pixmap {
            require(buffer.isDirect) { "Pixmap requires a direct ByteBuffer" }
            val addr = PixmapNative.nBufferAddress(buffer)
            return Pixmap(PixmapNative.nMake(info.nativePtr, addr, rowBytes), buffer)
        }

        internal fun wrapNative(
            ptr: Long,
            keepAlive: Any?,
        ): Pixmap = Pixmap(ptr, keepAlive)
    }
}

private object PixmapNative {
    init {
        NativeLoader.ensureLoaded()
    }

    external fun nBufferAddress(buffer: ByteBuffer): Long

    external fun nMake(
        infoPtr: Long,
        addr: Long,
        rowBytes: Long,
    ): Long

    external fun nRelease(ptr: Long)

    external fun nAddr(ptr: Long): Long

    external fun nRowBytes(ptr: Long): Long

    external fun nWidth(ptr: Long): Int

    external fun nHeight(ptr: Long): Int

    external fun nIsEmpty(ptr: Long): Boolean

    external fun nColorType(ptr: Long): Int

    external fun nAlphaType(ptr: Long): Int

    external fun nIsOpaque(ptr: Long): Boolean

    external fun nRowBytesAsPixels(ptr: Long): Int

    external fun nShiftPerPixel(ptr: Long): Int

    external fun nComputeByteSize(ptr: Long): Long

    external fun nGetColor(
        ptr: Long,
        x: Int,
        y: Int,
    ): Int

    external fun nGetAlphaf(
        ptr: Long,
        x: Int,
        y: Int,
    ): Float

    external fun nExtractSubset(
        ptr: Long,
        left: Int,
        top: Int,
        right: Int,
        bottom: Int,
    ): Long

    external fun nReadPixels(
        ptr: Long,
        dstPtr: Long,
        srcX: Int,
        srcY: Int,
    ): Boolean

    external fun nScalePixels(
        ptr: Long,
        dstPtr: Long,
        maxAniso: Int,
        useCubic: Boolean,
        cubicB: Float,
        cubicC: Float,
        filter: Int,
        mipmap: Int,
    ): Boolean
}

File: rust/skialin-jni/src/pixmap.rs
use jni::objects::JByteBuffer;
use jni::sys::{jboolean, jfloat, jint, jlong, jobject};
use jni::JNIEnv;

use skialin_core::{ImageInfo, Pixmap};

use crate::color_type::{alpha_type_to_ordinal, color_type_to_ordinal};
use crate::util::{borrow, borrow_mut, box_ptr, drop_ptr};

fn sampling_from_args(max_aniso: jint, use_cubic: jboolean, cubic_b: f32, cubic_c: f32, filter: jint, mipmap: jint) -> skialin_core::SamplingOptions {
    skialin_core::SamplingOptions {
        max_aniso,
        cubic: (use_cubic != 0).then_some((cubic_b, cubic_c)),
        filter: if filter == 1 { skialin_core::FilterMode::Linear } else { skialin_core::FilterMode::Nearest },
        mipmap: match mipmap {
            1 => skialin_core::MipmapMode::Nearest,
            2 => skialin_core::MipmapMode::Linear,
            _ => skialin_core::MipmapMode::None,
        },
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nBufferAddress(env: JNIEnv, _class: jni::objects::JClass, buffer: jobject) -> jlong {
    let buffer = unsafe { JByteBuffer::from_raw(buffer) };
    env.get_direct_buffer_address(&buffer).expect("get_direct_buffer_address") as jlong
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nMake(_env: JNIEnv, _class: jni::objects::JClass, info_ptr: jlong, addr: jlong, row_bytes: jlong) -> jlong {
    let info = unsafe { borrow::<ImageInfo>(info_ptr) };
    box_ptr(unsafe { Pixmap::new(info, addr as *const u8, row_bytes as usize) })
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nRelease(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) {
    unsafe { drop_ptr::<Pixmap>(ptr) };
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nAddr(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jlong {
    unsafe { borrow::<Pixmap>(ptr) }.addr() as jlong
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nRowBytes(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jlong {
    unsafe { borrow::<Pixmap>(ptr) }.row_bytes() as jlong
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nWidth(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jint {
    unsafe { borrow::<Pixmap>(ptr) }.width()
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nHeight(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jint {
    unsafe { borrow::<Pixmap>(ptr) }.height()
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nIsEmpty(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jboolean {
    unsafe { borrow::<Pixmap>(ptr) }.is_empty() as jboolean
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nColorType(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jint {
    color_type_to_ordinal(unsafe { borrow::<Pixmap>(ptr) }.color_type())
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nAlphaType(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jint {
    alpha_type_to_ordinal(unsafe { borrow::<Pixmap>(ptr) }.alpha_type())
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nIsOpaque(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jboolean {
    unsafe { borrow::<Pixmap>(ptr) }.is_opaque() as jboolean
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nRowBytesAsPixels(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jint {
    unsafe { borrow::<Pixmap>(ptr) }.row_bytes_as_pixels()
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nShiftPerPixel(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jint {
    unsafe { borrow::<Pixmap>(ptr) }.shift_per_pixel()
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nComputeByteSize(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong) -> jlong {
    unsafe { borrow::<Pixmap>(ptr) }.compute_byte_size() as jlong
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nGetColor(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong, x: jint, y: jint) -> jint {
    unsafe { borrow::<Pixmap>(ptr) }.get_color(x, y) as jint
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nGetAlphaf(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong, x: jint, y: jint) -> jfloat {
    unsafe { borrow::<Pixmap>(ptr) }.get_alphaf(x, y)
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nExtractSubset(
    _env: JNIEnv,
    _class: jni::objects::JClass,
    ptr: jlong,
    left: jint,
    top: jint,
    right: jint,
    bottom: jint,
) -> jlong {
    let area = skialin_core::IRect::new(left, top, right, bottom);
    match unsafe { borrow::<Pixmap>(ptr).extract_subset(area) } {
        Some(subset) => box_ptr(subset),
        None => 0,
    }
}

#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nReadPixels(_env: JNIEnv, _class: jni::objects::JClass, ptr: jlong, dst_ptr: jlong, src_x: jint, src_y: jint) -> jboolean {
    let dst = unsafe { borrow_mut::<Pixmap>(dst_ptr) };
    unsafe { borrow::<Pixmap>(ptr) }.read_pixels(dst, src_x, src_y) as jboolean
}

#[allow(clippy::too_many_arguments)]
#[no_mangle]
pub extern "system" fn Java_org_skialin_PixmapNative_nScalePixels(
    _env: JNIEnv,
    _class: jni::objects::JClass,
    ptr: jlong,
    dst_ptr: jlong,
    max_aniso: jint,
    use_cubic: jboolean,
    cubic_b: jfloat,
    cubic_c: jfloat,
    filter: jint,
    mipmap: jint,
) -> jboolean {
    let dst = unsafe { borrow_mut::<Pixmap>(dst_ptr) };
    let sampling = sampling_from_args(max_aniso, use_cubic, cubic_b, cubic_c, filter, mipmap);
    unsafe { borrow::<Pixmap>(ptr) }.scale_pixels(dst, sampling) as jboolean
}
